### PR TITLE
New entry go goroutines

### LINKED ISSUE #944: [Entry] Go: Goroutines
### Reason/inspiration (optional)

Please write a new Go concept entry about the goroutines that includes the following:

- A "Syntax" section with a \```pseudo code snippet.
- An "Example" section with a \```go snippet.
- A "Codebyte Example" section with a \```codebyte/golang snippet.

Please refer to the [entry templates](https://github.com/Codecademy/docs/blob/main/documentation/), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entry for Existing Topic

### Code of Conduct

- [X] I agree to follow Codecademy Doc's Code of Conduct

### For Maintainers

- [X] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by cslylla**
### Description

- Added new concept for Go: goroutines

- Closes #944 

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### Additional Information
I am using npm; I have node 18.12.
I was unable to run format:verify (prettier is not recognized)
I was unable to run lint:md (markdownlint is not recognized)
I was unable to run test (jest is not recognized)

### CODE REVIEW

**Comment by cslylla** [comment]
Once #1242 is completed, I would like to refer to it when the definition of functions is described.

File: content/go/concepts/goroutines/goroutines.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  - 'Concurrency'
+  - 'Functions'
+CatalogContent:
+  - 'learn-go'
+  - 'paths/back-end-engineer-career-path'
+---
+
+**Goroutines** are functions and methods that run concurrently with other functions and methods.
+
+## The difference between threads and goroutines
```
Comment (SSwiniarski) [inline content/go/concepts/goroutines/goroutines.md:19]: I'd get rid of the header and incorporate the following paragraph into the introduction.

File: content/go/concepts/goroutines/goroutines.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+```codebyte/golang
+package main
+import (
+  "fmt"
+  "time"
+)
+
+// define a function
+func print(text string) {
+
```
Comment (SSwiniarski) [inline content/go/concepts/goroutines/goroutines.md:77]:
```suggestion
```

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@cslylla I did a first review and left a couple of comments.

**Review by SSwiniarski — APPROVED** [review]
@cslylla Looks good for a second review.

File: content/go/concepts/goroutines/goroutines.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  - 'Concurrency'
+  - 'Functions'
+CatalogContent:
+  - 'learn-go'
+  - 'paths/back-end-engineer-career-path'
+---
+
+**Goroutines** are functions and methods that run concurrently with other functions and methods.
+
+Although goroutines share some similarities with threads, goroutines are not threads. Threads are different processes within a program that are running at the same time. Threads are created by the program; they are smaller and lighter processes and have their own flow of control. Running multiple processes simultaneously allows complex programs to run without holding up the execution of the rest of the program. Go supports multi-threading with goroutines. Goroutines are lightweight threads managed by the Go runtime.
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:19]:
```suggestion
Although goroutines share some similarities with threads, there are important differences that include the following:

- Threads depend on the hardware of the host computer's operating system, whereas goroutines do not.
- Compared with threads, goroutines are cost-effective and use less resources to function.
- Goroutines do not use thread-local storage and, thus, do not have a unique ID.

Comparatively, goroutines are light-weight versions of threads that operate within the context of the Go runtime.
```

File: content/go/concepts/goroutines/goroutines.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+## Syntax
+
+To start a goroutine, invoke a previously defined function or method with the `go` keyword:
+
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:24]:
```suggestion

```

File: content/go/concepts/goroutines/goroutines.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+```pseudo
+func myFunction(parameter) returnType {
+   //body of the function
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:27]:
```suggestion
   // function body
```

File: content/go/concepts/goroutines/goroutines.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+go myFunction()
+```
+
+In the code above the `myFunction()` function is defined. Then the function is invoked with the `go` keyword. `myFunction()` will run concurrently with the calling function.
+
+> **Note:** for any goroutines to run, the `main()` function should be running. When the `main()` function terminates, the program will be terminated, and no other goroutine will run.
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:35]:
```suggestion
A goroutine is started by invoking a previously defined function or method, `myFunction()`, with the `go` keyword.

> **Note:** For any goroutines to run, the `main()` function must defined and executed. When the `main()` function terminates, the program will be terminated, and no other goroutine will run.
```

File: content/go/concepts/goroutines/goroutines.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+## Example
+
+In the example below, the `myGoroutine()` function is defined to print out the text `This is my first goroutine`. Then in the `main()` function, `myGoroutine()` is called as a goroutine with the `go` keyword. Afterwards, the `main()` function is set to wait two seconds with the `time.Sleep()` function and then print out the text `This is the main function`.
+
+```go
+func myGoroutine() {
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:42]:
```suggestion
package main
import (
  "fmt"
  "time"
)

func myGoroutine() {  
```

File: content/go/concepts/goroutines/goroutines.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+go myFunction()
+```
+
+In the code above the `myFunction()` function is defined. Then the function is invoked with the `go` keyword. `myFunction()` will run concurrently with the calling function.
+
+> **Note:** for any goroutines to run, the `main()` function should be running. When the `main()` function terminates, the program will be terminated, and no other goroutine will run.
+
+## Example
+
+In the example below, the `myGoroutine()` function is defined to print out the text `This is my first goroutine`. Then in the `main()` function, `myGoroutine()` is called as a goroutine with the `go` keyword. Afterwards, the `main()` function is set to wait two seconds with the `time.Sleep()` function and then print out the text `This is the main function`.
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:39]:
```suggestion
In the following example, a goroutine is defined to print out text. Inside the `main()` function, `myGoroutine()` is called with the `go` keyword and causes a time delay of 2 seconds with `time.Sleep()` before printing the text:
```

File: content/go/concepts/goroutines/goroutines.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+```go
+func myGoroutine() {  
+    fmt.Println("This is my first goroutine")
+}
+
+func main() {  
+    go myGoroutine()
+    time.Sleep(2 * time.Second)
+    fmt.Println("This is the main function")
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:49]:
```suggestion
  go myGoroutine()
  time.Sleep(2 * time.Second)
  fmt.Println("This is the main function")
```

File: content/go/concepts/goroutines/goroutines.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+}
+```
+
+This example results in the following output:
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:53]:
```suggestion
This results in the following output:
```

File: content/go/concepts/goroutines/goroutines.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+```shell
+This is my first goroutine
+This is the main function
+```
+
+> **Note:** When a goroutine is started, the goroutine call returns immediately. The program does not wait for the goroutine to finish executing. After the goroutine call, the program continues the next line of code, and any return values from the goroutine are ignored. The `time.Sleep()` function keeps the `main()` function running while the `myGoroutine()` function executes. Without the two seconds waiting time in the `time.Sleep()` function, the second print statement could be immediately completed, and the `main()` function could terminate before the goroutine is completed.
+
+## Codebyte Example
+
+The example below defines a function called `print()` to print out a string, which is first called as a goroutine and then as a function:
```
Comment (Dusch4593) [inline content/go/concepts/goroutines/goroutines.md:64]:
```suggestion
The example below defines a `print()` function to print out a string, which is first called as a goroutine and then as a regular function:
```

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@cslylla Left some comments/suggestions for you. 😄

**Comment by cslylla** [comment]
@Dusch4593, thanks for the comments; they are really good suggestions. 🤓 I also changed a few words/spacing. Let me know if that is okay!

**Review by Dusch4593 — APPROVED** [review]
@cslylla LGTM! 👍🏻

Just removed some extra spacing in d34a7d78e20ae4fdf63b237e74f7b42498f21900 to get all the checks to pass. 😄

**Comment by Dusch4593** [comment]
Thank you for another wonderful contribution to Codecademy Docs, @cslylla! 🎉 🎊 🥳

The entry should now be live: https://www.codecademy.com/resources/docs/go/goroutines

Screen Shot 2023-01-04 at 12 06 36 PM

Looking forward to your next Docs contribution! 😄

**Comment by cslylla** [comment]
Hey @lucwhite, I have just sent you an email! 🤩

**Comment by Pavan-Nambi** [comment]
sorry for tht i messed up few things while reverting ignore that

**Comment by yangc95** [comment]
@Pavan-Nambi That's alright! Please let me know if you need any assistance 🙂

**Pull request merged**